enhancements/microshift/microshift-storage-version-migrator.md

```diff
@@ -0,0 +1,267 @@
+---
+title: microshift-storage-version-migrator
+authors:
+  - eggfoobar
+reviewers:
+  - "dhellmann"
+  - "sanchezl"
+  - "pmtk"
+approvers:
+  - "dhellmann"
+api-approvers:
+  - None
+creation-date: 2023-07-17
+last-updated: 2023-07-25
+tracking-link:
+  - https://issues.redhat.com/browse/USHIFT-1473
+---
+
+# Embedding Storage Version Migrator
+
+## Summary
+
+We have embedded the
+[Migrator](https://github.com/openshift/kubernetes-kube-storage-version-migrator/tree/master/cmd/migrator)
+and
+[Trigger](https://github.com/openshift/kubernetes-kube-storage-version-migrator/tree/master/cmd/trigger)
+controllers in MicroShift to assure successive upgrades to MicroShift do not
+cause a situation where an API no longer exists for resource versions stored in
+ETCD.
+
+## Motivation
+
+We wanted to provide support in MicroShift for the same solution in OpenShift
+for automatic storage version migrations (i.e. v1alpha1 -> v1beta1 -> v1). The
+current design in OpenShift revolves around running the [Migrator
+Operator](https://github.com/openshift/cluster-kube-storage-version-migrator-operator)
+which only deploys the Migrator controller. Migrations are then left up to the
+Operator owners to kick off when changes happen to their respective APIs.
+OpenShift also uses the Migrator to help when encryption keys get rotated for
+ETCD, this allows OpenShift to re-write the objects to ETCD after an encryption
+change.
+
+In MicroShift we don't run the full scale of OpenShift so we can trigger
+migrations on start up to make sure resources are updated for future upgrades.
+Furthermore, things will continue to run smoothly when users install an Operator
+that might run with the assumption that a Migrator is running in the cluster and
+thus try and kick off a migration for it's resource.
+
+### User Stories
+
+As a user, I want to upgrade MicroShift with out worrying about the storage
+state of the resources already applied to the cluster.
+
+### Goals
+
+- On start up, MicroShift should run the Migrator controller as a service
+- On start up, MicroShift should run the Trigger controller as a service
+- MicroShift, should keep the resources updated to the Preferred Server Resource
+
+### Non-Goals
+
+- The Migrator encryption use case is not supported by MicroShift since it does
+  not run with etcd encryption, instead we support the use of on disk encryption
```

Comment:
If MicroShift is not supporting etcd encryption, it seems pre-mature to add the migrator at this point. Until upstream figures out migration, they will continue to support the old storage versions, even for APIs that have long been removed.

Reply:
@sanchezl does that apply to CRDs, too?

Reply:
No. If a CRD's versions are not managed carefully, (using the `served` and `storage` properties in the CRD), this could be a problem. 

For OpenShift provided CRDs we typically only make "compatible" changes and try to avoid actually changing the version. Otherwise we need to leave the older versions around in the CRD with `served:false` set.

As for customer CRDs, we leave that up to them to manage.

Reply:
How involved are we in the upstream conversations about how storage migration should work?

@deads2k you've previously indicated we should have the storage migration component, so I'm interested in your thoughts on this thread, too.